src/libstore/realisation.hh

```diff
@@ -28,6 +28,14 @@ struct Realisation {
 
     StringSet signatures;
 
+    /**
+     * The realisations that are required for the current one to be valid.
+     *
+     * When importing this realisation, the store will first check that all its
+     * dependencies exist, and map to the correct output path
+     */
+    std::map<DrvOutput, StorePath> dependentRealisations;
```

Comment:
I'm not sure I am comfortable referring to just the store path and not the entire realization. The whole point of this is coherence, and so we don't want just *any* realization connecting those two, but a coherent one. Similarly, the signatures ought to form some sort of Merkel dag.

Reply:
If we do that, I would change `Realization` to not have the `DrvOutput`, and call the original `RealizationWithKey` or similar. This avoids denormalization here and in `DrvOutputs`. (Actually, I think this should just reuse `DrvOutputs`.)

Reply:
In further thought, it should just be a regular content addressed (excluding signatures and other non-normative metadata) Merkle DAG.

We would store content addresses for each realization in the db. The map values can be (keyed or unkeyed, doesn't matter) realization *hashes*, not store paths. That keeps fetching from being any deeper than it is today, but still gives us the nice Merkel DAG guarantees.

It's content addressing all the way down!

Reply:
> I'm not sure I am comfortable referring to just the store path and not the entire realization. The whole point of this is coherence, and so we don't want just any realization connecting those two, but a coherent one.

Maybe I’m missing something, but for me “coherence” is just about having the same store paths everywhere. The rest of the realisations is either metadata that shouldn’t come into play here (signatures for example), or stuff that’s implied by the output path (the transitive coherence, because if two CA realisations have the same output path, then they are bound to be coherent)

Also, I didn’t make it contain the full realisation, because that would have implied that each realisation has to *physically* (for a rather broad notion of physically) contain its whole dependency tree, which would make things awfully slow and redundant.

> In further thought, it should just be a regular content addressed (excluding signatures and other non-normative metadata) Merkle DAG.

Mh, not sure I follow you here

Reply:
For reference, we discused this and decided it should be fine.